<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.1" />
<title>seqsearch.databases.foraminifera API documentation</title>
<meta name="description" content="Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>seqsearch.databases.foraminifera</code></h1>
</header>
<section id="section-intro">
<p>Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">#!/usr/bin/env python3
# -*- coding: utf-8 -*-

&#34;&#34;&#34;
Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio
&#34;&#34;&#34;

# Built-in modules #
import os

# First party modules #
from seqsearch.databases import Database
from fasta import FASTA
from autopaths.auto_paths import AutoPaths
from autopaths.file_path import FilePath

# Third party modules #

# Constants #
home = os.environ.get(&#39;HOME&#39;, &#39;~&#39;) + &#39;/&#39;

###############################################################################
class Foraminifera(Database):
    &#34;&#34;&#34;
    This is a custom database containing exclusively Foraminifera sequences.

    https://genev.unige.ch/research/laboratory/Jan-Pawlowski

    You should place the file &#34;foram_db_cor.fasta&#34; in:  ~/databases/foraminifera/
    Then you can run this:

            from seqsearch.databases.foraminifera import foraminifera
            foraminifera.process()
            print foraminifera.tax_depth_freq

    &#34;&#34;&#34;

    short_name = &#34;foraminifera&#34;
    long_name  = &#39;The custom made Foraminifera database as received by email on 7th April 2017&#39;

    all_paths = &#34;&#34;&#34;
    /foram_db_cor.fasta
    /foram_mothur.fasta
    /foram_mothur.tax
    &#34;&#34;&#34;

    @property
    def rank_names(self):
        &#34;&#34;&#34;The names of the ranks. Total 9 ranks.&#34;&#34;&#34;
        return [&#39;Domain&#39;,   # 0
                &#39;Kingdom&#39;,  # 1
                &#39;Phylum&#39;,   # 2
                &#39;Class&#39;,    # 3
                &#39;Order&#39;,    # 4
                &#39;Family&#39;,   # 5
                &#39;Tribe&#39;,    # 6
                &#39;Genus&#39;,    # 7
                &#39;Species&#39;]  # 8

    def __init__(self, base_dir=None):
        # Base directory #
        if base_dir is None: base_dir = home
        self.base_dir = base_dir + &#39;databases/&#39; + self.short_name + &#39;/&#39;
        self.p        = AutoPaths(self.base_dir, self.all_paths)
        # The results #
        self.alignment = FASTA(self.p.mothur_fasta)
        self.taxonomy  = FilePath(self.p.mothur_tax)
        # The part that mothur will use for naming files #
        self.nickname = &#34;foram_mothur&#34;

    def process(self):
        # The file that was received by email without documentation #
        raw = FASTA(self.p.cor)
        # Open files #
        self.alignment.create()
        self.taxonomy.create()
        # Loop #
        for seq in raw:
            # Parse #
            name = seq.id[11:].split(&#39;|&#39;)
            num  = name.pop(0)
            # Check #
            for x in name: assert &#39;;&#39; not in x
            for x in name: assert &#39;\t&#39; not in x
            # Make ranks #
            ranks = [&#39;Eukaryota&#39;                       , # 0 Domain
                     &#39;Rhizaria&#39;                        , # 1 Kingdom
                     &#39;Foraminifera&#39;                    , # 2 Phylum
                     name[0]                           , # 3 Class
                     name[1]                           , # 4 Order
                     name[2]                           , # 5 Family
                     name[3]                           , # 6 Tribe
                     name[4]                           , # 7 Genus
                     name[5]]                            # 8 Species
            # The taxonomy string #
            tax_line = &#39;;&#39;.join(ranks)
            # Add sequence to the new fasta file #
            self.alignment.add_str(str(seq.seq), name=&#34;foram&#34; + num)
            # Add the taxonomy to the tax file #
            self.taxonomy.add_str(&#34;foram&#34; + num + &#39;\t&#39; + tax_line + &#39;\n&#39;)
        # Close files #
        self.alignment.close()
        self.taxonomy.close()

###############################################################################
foraminifera = Foraminifera()</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="seqsearch.databases.foraminifera.Foraminifera"><code class="flex name class">
<span>class <span class="ident">Foraminifera</span></span>
<span>(</span><span>base_dir=None)</span>
</code></dt>
<dd>
<div class="desc"><p>This is a custom database containing exclusively Foraminifera sequences.</p>
<p><a href="https://genev.unige.ch/research/laboratory/Jan-Pawlowski">https://genev.unige.ch/research/laboratory/Jan-Pawlowski</a></p>
<p>You should place the file "foram_db_cor.fasta" in:
~/databases/foraminifera/
Then you can run this:</p>
<pre><code>    from seqsearch.databases.foraminifera import foraminifera
    foraminifera.process()
    print foraminifera.tax_depth_freq
</code></pre></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class Foraminifera(Database):
    &#34;&#34;&#34;
    This is a custom database containing exclusively Foraminifera sequences.

    https://genev.unige.ch/research/laboratory/Jan-Pawlowski

    You should place the file &#34;foram_db_cor.fasta&#34; in:  ~/databases/foraminifera/
    Then you can run this:

            from seqsearch.databases.foraminifera import foraminifera
            foraminifera.process()
            print foraminifera.tax_depth_freq

    &#34;&#34;&#34;

    short_name = &#34;foraminifera&#34;
    long_name  = &#39;The custom made Foraminifera database as received by email on 7th April 2017&#39;

    all_paths = &#34;&#34;&#34;
    /foram_db_cor.fasta
    /foram_mothur.fasta
    /foram_mothur.tax
    &#34;&#34;&#34;

    @property
    def rank_names(self):
        &#34;&#34;&#34;The names of the ranks. Total 9 ranks.&#34;&#34;&#34;
        return [&#39;Domain&#39;,   # 0
                &#39;Kingdom&#39;,  # 1
                &#39;Phylum&#39;,   # 2
                &#39;Class&#39;,    # 3
                &#39;Order&#39;,    # 4
                &#39;Family&#39;,   # 5
                &#39;Tribe&#39;,    # 6
                &#39;Genus&#39;,    # 7
                &#39;Species&#39;]  # 8

    def __init__(self, base_dir=None):
        # Base directory #
        if base_dir is None: base_dir = home
        self.base_dir = base_dir + &#39;databases/&#39; + self.short_name + &#39;/&#39;
        self.p        = AutoPaths(self.base_dir, self.all_paths)
        # The results #
        self.alignment = FASTA(self.p.mothur_fasta)
        self.taxonomy  = FilePath(self.p.mothur_tax)
        # The part that mothur will use for naming files #
        self.nickname = &#34;foram_mothur&#34;

    def process(self):
        # The file that was received by email without documentation #
        raw = FASTA(self.p.cor)
        # Open files #
        self.alignment.create()
        self.taxonomy.create()
        # Loop #
        for seq in raw:
            # Parse #
            name = seq.id[11:].split(&#39;|&#39;)
            num  = name.pop(0)
            # Check #
            for x in name: assert &#39;;&#39; not in x
            for x in name: assert &#39;\t&#39; not in x
            # Make ranks #
            ranks = [&#39;Eukaryota&#39;                       , # 0 Domain
                     &#39;Rhizaria&#39;                        , # 1 Kingdom
                     &#39;Foraminifera&#39;                    , # 2 Phylum
                     name[0]                           , # 3 Class
                     name[1]                           , # 4 Order
                     name[2]                           , # 5 Family
                     name[3]                           , # 6 Tribe
                     name[4]                           , # 7 Genus
                     name[5]]                            # 8 Species
            # The taxonomy string #
            tax_line = &#39;;&#39;.join(ranks)
            # Add sequence to the new fasta file #
            self.alignment.add_str(str(seq.seq), name=&#34;foram&#34; + num)
            # Add the taxonomy to the tax file #
            self.taxonomy.add_str(&#34;foram&#34; + num + &#39;\t&#39; + tax_line + &#39;\n&#39;)
        # Close files #
        self.alignment.close()
        self.taxonomy.close()</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="seqsearch.databases.Database" href="index.html#seqsearch.databases.Database">Database</a></li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="seqsearch.databases.foraminifera.Foraminifera.all_paths"><code class="name">var <span class="ident">all_paths</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="seqsearch.databases.foraminifera.Foraminifera.long_name"><code class="name">var <span class="ident">long_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="seqsearch.databases.foraminifera.Foraminifera.short_name"><code class="name">var <span class="ident">short_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Instance variables</h3>
<dl>
<dt id="seqsearch.databases.foraminifera.Foraminifera.rank_names"><code class="name">var <span class="ident">rank_names</span></code></dt>
<dd>
<div class="desc"><p>The names of the ranks. Total 9 ranks.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def rank_names(self):
    &#34;&#34;&#34;The names of the ranks. Total 9 ranks.&#34;&#34;&#34;
    return [&#39;Domain&#39;,   # 0
            &#39;Kingdom&#39;,  # 1
            &#39;Phylum&#39;,   # 2
            &#39;Class&#39;,    # 3
            &#39;Order&#39;,    # 4
            &#39;Family&#39;,   # 5
            &#39;Tribe&#39;,    # 6
            &#39;Genus&#39;,    # 7
            &#39;Species&#39;]  # 8</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="seqsearch.databases.foraminifera.Foraminifera.process"><code class="name flex">
<span>def <span class="ident">process</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def process(self):
    # The file that was received by email without documentation #
    raw = FASTA(self.p.cor)
    # Open files #
    self.alignment.create()
    self.taxonomy.create()
    # Loop #
    for seq in raw:
        # Parse #
        name = seq.id[11:].split(&#39;|&#39;)
        num  = name.pop(0)
        # Check #
        for x in name: assert &#39;;&#39; not in x
        for x in name: assert &#39;\t&#39; not in x
        # Make ranks #
        ranks = [&#39;Eukaryota&#39;                       , # 0 Domain
                 &#39;Rhizaria&#39;                        , # 1 Kingdom
                 &#39;Foraminifera&#39;                    , # 2 Phylum
                 name[0]                           , # 3 Class
                 name[1]                           , # 4 Order
                 name[2]                           , # 5 Family
                 name[3]                           , # 6 Tribe
                 name[4]                           , # 7 Genus
                 name[5]]                            # 8 Species
        # The taxonomy string #
        tax_line = &#39;;&#39;.join(ranks)
        # Add sequence to the new fasta file #
        self.alignment.add_str(str(seq.seq), name=&#34;foram&#34; + num)
        # Add the taxonomy to the tax file #
        self.taxonomy.add_str(&#34;foram&#34; + num + &#39;\t&#39; + tax_line + &#39;\n&#39;)
    # Close files #
    self.alignment.close()
    self.taxonomy.close()</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="seqsearch.databases.Database" href="index.html#seqsearch.databases.Database">Database</a></b></code>:
<ul class="hlist">
<li><code><a title="seqsearch.databases.Database.blast_db" href="index.html#seqsearch.databases.Database.blast_db">blast_db</a></code></li>
<li><code><a title="seqsearch.databases.Database.download" href="index.html#seqsearch.databases.Database.download">download</a></code></li>
<li><code><a title="seqsearch.databases.Database.files_remaining" href="index.html#seqsearch.databases.Database.files_remaining">files_remaining</a></code></li>
<li><code><a title="seqsearch.databases.Database.files_to_retrieve" href="index.html#seqsearch.databases.Database.files_to_retrieve">files_to_retrieve</a></code></li>
<li><code><a title="seqsearch.databases.Database.ftp" href="index.html#seqsearch.databases.Database.ftp">ftp</a></code></li>
<li><code><a title="seqsearch.databases.Database.raw_files" href="index.html#seqsearch.databases.Database.raw_files">raw_files</a></code></li>
<li><code><a title="seqsearch.databases.Database.sequences" href="index.html#seqsearch.databases.Database.sequences">sequences</a></code></li>
<li><code><a title="seqsearch.databases.Database.tax_depth_freq" href="index.html#seqsearch.databases.Database.tax_depth_freq">tax_depth_freq</a></code></li>
<li><code><a title="seqsearch.databases.Database.ungzip" href="index.html#seqsearch.databases.Database.ungzip">ungzip</a></code></li>
<li><code><a title="seqsearch.databases.Database.untargz" href="index.html#seqsearch.databases.Database.untargz">untargz</a></code></li>
</ul>
</li>
</ul>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="seqsearch.databases" href="index.html">seqsearch.databases</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="seqsearch.databases.foraminifera.Foraminifera" href="#seqsearch.databases.foraminifera.Foraminifera">Foraminifera</a></code></h4>
<ul class="">
<li><code><a title="seqsearch.databases.foraminifera.Foraminifera.all_paths" href="#seqsearch.databases.foraminifera.Foraminifera.all_paths">all_paths</a></code></li>
<li><code><a title="seqsearch.databases.foraminifera.Foraminifera.long_name" href="#seqsearch.databases.foraminifera.Foraminifera.long_name">long_name</a></code></li>
<li><code><a title="seqsearch.databases.foraminifera.Foraminifera.process" href="#seqsearch.databases.foraminifera.Foraminifera.process">process</a></code></li>
<li><code><a title="seqsearch.databases.foraminifera.Foraminifera.rank_names" href="#seqsearch.databases.foraminifera.Foraminifera.rank_names">rank_names</a></code></li>
<li><code><a title="seqsearch.databases.foraminifera.Foraminifera.short_name" href="#seqsearch.databases.foraminifera.Foraminifera.short_name">short_name</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.1</a>.</p>
</footer>
</body>
</html>