<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.2" />
<title>seqsearch.databases.silva API documentation</title>
<meta name="description" content="Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>seqsearch.databases.silva</code></h1>
</header>
<section id="section-intro">
<p>Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">#!/usr/bin/env python3
# -*- coding: utf-8 -*-

&#34;&#34;&#34;
Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio
&#34;&#34;&#34;

# Built-in modules #
import os

# First party modules #
from seqsearch.databases import Database
from fasta import FASTA
from autopaths.auto_paths import AutoPaths

# Third party modules #

# Constants #
home = os.environ.get(&#39;HOME&#39;, &#39;~&#39;) + &#39;/&#39;

###############################################################################
class Silva(Database):
    &#34;&#34;&#34;
    SILVA provides comprehensive, quality checked and regularly updated
    datasets of aligned small (16S/18S, SSU) and large subunit (23S/28S, LSU)
    ribosomal RNA (rRNA) sequences for all three domains of life
    (Bacteria, Archaea and Eukarya).
    SILVA are the official databases of the software package ARB.

    https://www.arb-silva.de

    To install:
        from seqsearch.databases.silva import silva
        silva.download()
        silva.unzip()

    It will put it in ~/databases/silva_xxx/
    &#34;&#34;&#34;

    view_url   = &#34;https://www.arb-silva.de/no_cache/download/archive/&#34;
    base_url   = &#34;https://www.arb-silva.de/fileadmin/silva_databases/&#34;
    short_name = &#34;silva&#34;

    all_paths = &#34;&#34;&#34;
    /test.txt
    &#34;&#34;&#34;

    def __init__(self, version, seq_type, base_dir=None):
        # Attributes #
        self.version    = version
        self.seq_type   = seq_type
        self.short_name = self.short_name + &#34;_&#34; + self.version
        # Base directory #
        if base_dir is None: base_dir = home
        self.base_dir = base_dir + &#39;databases/&#39; + self.short_name + &#39;/&#39;
        self.p        = AutoPaths(self.base_dir, self.all_paths)
        # URL #
        self.url  = &#34;release_%s/Exports/&#34;  % self.version
        # The database #
        self.nr99_name = &#34;SILVA_%s_SSURef_Nr99_tax_silva.fasta.gz&#34; % self.version
        self.nr99_dest = FASTA(self.base_dir + self.nr99_name)
        self.nr99      = FASTA(self.base_dir + self.nr99_name[:-3])
        # The alignment #
        self.aligned_name = &#34;SILVA_%s_SSURef_Nr99_tax_silva_full_align_trunc.fasta.gz&#34; % self.version
        self.aligned_dest = FASTA(self.base_dir + self.aligned_name)
        self.aligned      = FASTA(self.base_dir + self.aligned_name[:-3])

    def download(self):
        self.nr99_dest.directory.create(safe=True)
        self.nr99_dest.remove(safe=True)
        self.aligned_dest.remove(safe=True)
        import wget
        print(&#34;\nDownloading&#34;, self.base_url + self.url + self.nr99_name)
        wget.download(self.base_url + self.url + self.nr99_name,    out=self.nr99_dest.path)
        print(&#34;\nDownloading&#34;, self.base_url + self.url + self.aligned_name)
        wget.download(self.base_url + self.url + self.aligned_name, out=self.aligned_dest.path)

    def unzip(self):
        self.nr99_dest.ungzip_to(self.nr99)
        self.nr99.permissions.only_readable()
        self.aligned_dest.ungzip_to(self.aligned)
        self.aligned.permissions.only_readable()

###############################################################################
silva = Silva(&#34;128&#34;, &#34;nucl&#34;)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="seqsearch.databases.silva.Silva"><code class="flex name class">
<span>class <span class="ident">Silva</span></span>
<span>(</span><span>version, seq_type, base_dir=None)</span>
</code></dt>
<dd>
<div class="desc"><p>SILVA provides comprehensive, quality checked and regularly updated
datasets of aligned small (16S/18S, SSU) and large subunit (23S/28S, LSU)
ribosomal RNA (rRNA) sequences for all three domains of life
(Bacteria, Archaea and Eukarya).
SILVA are the official databases of the software package ARB.</p>
<p><a href="https://www.arb-silva.de">https://www.arb-silva.de</a></p>
<p>To install:
from seqsearch.databases.silva import silva
silva.download()
silva.unzip()</p>
<p>It will put it in ~/databases/silva_xxx/</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class Silva(Database):
    &#34;&#34;&#34;
    SILVA provides comprehensive, quality checked and regularly updated
    datasets of aligned small (16S/18S, SSU) and large subunit (23S/28S, LSU)
    ribosomal RNA (rRNA) sequences for all three domains of life
    (Bacteria, Archaea and Eukarya).
    SILVA are the official databases of the software package ARB.

    https://www.arb-silva.de

    To install:
        from seqsearch.databases.silva import silva
        silva.download()
        silva.unzip()

    It will put it in ~/databases/silva_xxx/
    &#34;&#34;&#34;

    view_url   = &#34;https://www.arb-silva.de/no_cache/download/archive/&#34;
    base_url   = &#34;https://www.arb-silva.de/fileadmin/silva_databases/&#34;
    short_name = &#34;silva&#34;

    all_paths = &#34;&#34;&#34;
    /test.txt
    &#34;&#34;&#34;

    def __init__(self, version, seq_type, base_dir=None):
        # Attributes #
        self.version    = version
        self.seq_type   = seq_type
        self.short_name = self.short_name + &#34;_&#34; + self.version
        # Base directory #
        if base_dir is None: base_dir = home
        self.base_dir = base_dir + &#39;databases/&#39; + self.short_name + &#39;/&#39;
        self.p        = AutoPaths(self.base_dir, self.all_paths)
        # URL #
        self.url  = &#34;release_%s/Exports/&#34;  % self.version
        # The database #
        self.nr99_name = &#34;SILVA_%s_SSURef_Nr99_tax_silva.fasta.gz&#34; % self.version
        self.nr99_dest = FASTA(self.base_dir + self.nr99_name)
        self.nr99      = FASTA(self.base_dir + self.nr99_name[:-3])
        # The alignment #
        self.aligned_name = &#34;SILVA_%s_SSURef_Nr99_tax_silva_full_align_trunc.fasta.gz&#34; % self.version
        self.aligned_dest = FASTA(self.base_dir + self.aligned_name)
        self.aligned      = FASTA(self.base_dir + self.aligned_name[:-3])

    def download(self):
        self.nr99_dest.directory.create(safe=True)
        self.nr99_dest.remove(safe=True)
        self.aligned_dest.remove(safe=True)
        import wget
        print(&#34;\nDownloading&#34;, self.base_url + self.url + self.nr99_name)
        wget.download(self.base_url + self.url + self.nr99_name,    out=self.nr99_dest.path)
        print(&#34;\nDownloading&#34;, self.base_url + self.url + self.aligned_name)
        wget.download(self.base_url + self.url + self.aligned_name, out=self.aligned_dest.path)

    def unzip(self):
        self.nr99_dest.ungzip_to(self.nr99)
        self.nr99.permissions.only_readable()
        self.aligned_dest.ungzip_to(self.aligned)
        self.aligned.permissions.only_readable()</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="seqsearch.databases.Database" href="index.html#seqsearch.databases.Database">Database</a></li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="seqsearch.databases.silva.Silva.all_paths"><code class="name">var <span class="ident">all_paths</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="seqsearch.databases.silva.Silva.base_url"><code class="name">var <span class="ident">base_url</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="seqsearch.databases.silva.Silva.short_name"><code class="name">var <span class="ident">short_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="seqsearch.databases.silva.Silva.view_url"><code class="name">var <span class="ident">view_url</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="seqsearch.databases.silva.Silva.unzip"><code class="name flex">
<span>def <span class="ident">unzip</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def unzip(self):
    self.nr99_dest.ungzip_to(self.nr99)
    self.nr99.permissions.only_readable()
    self.aligned_dest.ungzip_to(self.aligned)
    self.aligned.permissions.only_readable()</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="seqsearch.databases.Database" href="index.html#seqsearch.databases.Database">Database</a></b></code>:
<ul class="hlist">
<li><code><a title="seqsearch.databases.Database.blast_db" href="index.html#seqsearch.databases.Database.blast_db">blast_db</a></code></li>
<li><code><a title="seqsearch.databases.Database.download" href="index.html#seqsearch.databases.Database.download">download</a></code></li>
<li><code><a title="seqsearch.databases.Database.files_remaining" href="index.html#seqsearch.databases.Database.files_remaining">files_remaining</a></code></li>
<li><code><a title="seqsearch.databases.Database.files_to_retrieve" href="index.html#seqsearch.databases.Database.files_to_retrieve">files_to_retrieve</a></code></li>
<li><code><a title="seqsearch.databases.Database.ftp" href="index.html#seqsearch.databases.Database.ftp">ftp</a></code></li>
<li><code><a title="seqsearch.databases.Database.raw_files" href="index.html#seqsearch.databases.Database.raw_files">raw_files</a></code></li>
<li><code><a title="seqsearch.databases.Database.sequences" href="index.html#seqsearch.databases.Database.sequences">sequences</a></code></li>
<li><code><a title="seqsearch.databases.Database.tax_depth_freq" href="index.html#seqsearch.databases.Database.tax_depth_freq">tax_depth_freq</a></code></li>
<li><code><a title="seqsearch.databases.Database.ungzip" href="index.html#seqsearch.databases.Database.ungzip">ungzip</a></code></li>
<li><code><a title="seqsearch.databases.Database.untargz" href="index.html#seqsearch.databases.Database.untargz">untargz</a></code></li>
</ul>
</li>
</ul>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="seqsearch.databases" href="index.html">seqsearch.databases</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="seqsearch.databases.silva.Silva" href="#seqsearch.databases.silva.Silva">Silva</a></code></h4>
<ul class="">
<li><code><a title="seqsearch.databases.silva.Silva.all_paths" href="#seqsearch.databases.silva.Silva.all_paths">all_paths</a></code></li>
<li><code><a title="seqsearch.databases.silva.Silva.base_url" href="#seqsearch.databases.silva.Silva.base_url">base_url</a></code></li>
<li><code><a title="seqsearch.databases.silva.Silva.short_name" href="#seqsearch.databases.silva.Silva.short_name">short_name</a></code></li>
<li><code><a title="seqsearch.databases.silva.Silva.unzip" href="#seqsearch.databases.silva.Silva.unzip">unzip</a></code></li>
<li><code><a title="seqsearch.databases.silva.Silva.view_url" href="#seqsearch.databases.silva.Silva.view_url">view_url</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.2</a>.</p>
</footer>
</body>
</html>